File: HashMap/149. Max Points on a Line/Solution.py
# Definition for a point.
# class Point(object):
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b

class Solution(object):
    def maxPoints(self, points):
        """
        :type points: List[Point]
        :rtype: int
        """
        n = len(points)
        if n<=1:
            return n
        result=0
        for i in range(n):
            pivot = points[i]
            record = {}
            sameX=0
            sameP=1
            maxPoint=0
            for j in range(i+1, n):
                x = pivot.x-points[j].x
                y = pivot.y-points[j].y
                if x==0 and y==0:
                    sameP+=1
                elif x==0:
                    sameX+=1
                else:
                    gcd = self.computeGCD(x, y)
                    if gcd!=0:
                        x /= gcd
                        y /= gcd
                    slope = (y,x)
                    record[slope] = record.get(slope, 0)+1
                    maxPoint = max(maxPoint, record[slope])
            maxPoint = max(maxPoint+sameP, sameX+sameP)
            result = max(result, maxPoint)
        return result
                    
    def computeGCD(self, x, y):
        while y!=0:
            tmp = y
            y = x%y
            x = tmp
        return x
        

File: HashMap/149. Max Points on a Line/Solution.java
/**
 * Definition for a point.
 * class Point {
 *     int x;
 *     int y;
 *     Point() { x = 0; y = 0; }
 *     Point(int a, int b) { x = a; y = b; }
 * }
 */
class Solution {
    //1. Map<slope, # of points>
    //2. for each point, find max point in the line
    //3. be careful with point align as veritical line
    public int maxPoints(Point[] points) {
        int n = points.length;
        if (n<=1){
            return n;
        }
        int max = 0;
        for (int i=0; i<n; i++){
            Point pivot = points[i];
            Map<String, Integer> map = new HashMap<>();
            int sameX=0, samePoint=1, maxPoints=0;
            for (int j=i+1; j<n; j++){
                if (points[j].x==pivot.x&&points[j].y==pivot.y){
                    samePoint++;
                }else if (points[j].x==pivot.x){
                    sameX++;
                }else{
                    //get their most common divisor, x must not be zero
                    int x = pivot.x-points[j].x;
                    int y = pivot.y-points[j].y;
                    int divisor = gcd(y,x);
                    String slope = (y/divisor)+"/"+(x/divisor);
                    map.put(slope, map.getOrDefault(slope, 0)+1);
                    int p = map.get(slope);
                    maxPoints = Math.max(maxPoints, p);
                }
            }
            maxPoints = Math.max(maxPoints, sameX);
            max = Math.max(max, maxPoints+samePoint);
        }
        return max;
    }
    private int gcd(int a, int b){
        //a is divisale by b, return b
        while (a%b!=0){
            int tmp=b;
            b = a%b;
            a = tmp;
        }
        return b;
    }
}